_posts/2023-11-10-automated-test-generation-chess-puzzles-with-dafny.markdown

```diff
@@ -0,0 +1,288 @@
+---
+layout: post
+title:  "Automated Test Generation: Chess Puzzles with Dafny"
+date:   2023-11-10 11:00:00 -0500
+author: Sasha Fedchin
+---
+Given all the incredible examples that showcase how powerful Dafny is (did you know that part of Dafny is written in Dafny?), you might think that verification can fully replace testing. And in many cases it does! However if, like a number of industrial Dafny users, you want to integrate Dafny code with existing codebases, you will face challenges that verification alone might not solve:
+
+- Perhaps your code is actually a reference implementation for another program and you want to test the equivalence between the two.
+- You might be calling out to external libraries, which are not themselves verified, and would like to test them by calling your Dafny code.
+- You have written high-level specifications for your code and are about to embark on a proof but want to have some initial assurance that your yet-to-be-verified implementation does conform to these specifications. 
+
+In this blog post, I show how Dafny’s built-in automated test generation can be useful in these situations. In the example I use, Dafny has to position pieces on a chess board in such a way that certain constraints are satisfied. This is exactly what I want to illustrate: chess boards will be generated tests, and constraints will be the program we want to test. A brute force enumeration or fuzzing might not be up to the task here, since the number of ways in which you can arrange pieces on the board is astronomically high, even under the constraints I add to the problem. Dafny, however, can use the verifier to generate tests and is, therefore, much more efficient in finding solutions. 
```

Comment:
> Dafny has to position pieces on a chess board

Does Dafny, the language, have to position pieces on a chess board? How can a language do that?

> This is exactly what I want to illustrate: chess boards will be generated tests, and constraints will be the program we want to test.

Don't you want to use the example to illustrate a higher level concept?

Consider:

"I will use Dafny's test generation feature to search for chess board states that satisfy particular interesting conditions, such as the white king being checked. The generated tests will contain the board states, while the interesting conditions are defined by the Dafny program under test"

Reply:
I now specifically refer to Dafny's test generation toolkit when describing this functionality and have incorporated the paraphrase you suggest!